File: src/Automator.Web/Models/AuditLog.cs
namespace Automator.Web.Models;

public class AuditLog
{
    public long Id { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
    public string? Username { get; set; }
    public string Action { get; set; } = string.Empty;
    public string? Resource { get; set; }
    public string? Details { get; set; }
}

File: automator/app/Models/AuditLog.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUlids;
use Illuminate\Database\Eloquent\Model;

class AuditLog extends Model
{
    use HasUlids;

    const UPDATED_AT = null;

    protected $fillable = ['username', 'action', 'resource', 'details'];

    public static function record(string $action, ?string $resource = null, ?string $details = null, ?string $username = null): void
    {
        static::create([
            'username' => $username ?? auth()->user()?->username,
            'action' => $action,
            'resource' => $resource,
            'details' => $details,
        ]);
    }
}
